lib/smartcar_error.rb

```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+# Custom SmartcarError class to represent errors from Smartcar APIs.
+class SmartcarError < StandardError
+  attr_reader :error, :code, :status_code, :request_id, :type, :description, :doc_url, :resolution, :detail
+
+  def initialize(status, body, headers)
+    @status_code = status
+    if body.is_a?(String)
+      super(body)
+      @request_id = headers['sc-request-id']
+      return
+    elsif body[:type] && body[:description]
+      super("#{body[:type]}:#{body[:code]} - #{body[:description]}")
+    else
+      super(body[:message] || 'Unknown error')
```

Comment:
need handling of oauth error's `error_description`

Reply:
There was a completely different problem here. THe oauth errors were not being coerced to smartcar error. I have added that in . The gem does convert 'error_Description' to 'description' but to keep the code logix similar i just used the raw body which has 'error_Description'